man/coredump-ng.1
.TH COREDUMP-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
coredump-ng: display information about ELF core dumps.

.SH SYNOPSIS
coredump-ng [\fB\-H\fR|\fB\-\-help\fR]
            [\fB\-v\fR|\fB\-\-version\fR]
            [\fB\-A\fR|\fB\-\-arch-specific\fR]
            [\fB\-I\fR|\fB\-\-histogram\fR]
            [\fB\-S\fR|\fB\-\-section-headers\fR]
            [\fB\-V\fR|\fB\-\-version-info\fR]
            [\fB\-a\fR|\fB\-\-all\fR]
            [\fB\-d\fR|\fB\-\-dynamic\fR]
            [\fB\-e\fR|\fB\-\-headers\fR]
            [\fB\-g\fR|\fB\-\-section-groups\fR]
            [\fB\-t\fR|\fB\-\-section-details\fR]
            [\fB\-h\fR|\fB\-\-file-header\fR]
            [\fB\-l\fR|\fB\-\-program-headers\fR]
            [\fB\-n\fR|\fB\-\-notes\fR]
            [\fB\-r\fR|\fB\-\-relocs\fR]
            [\fB\-s\fR|\fB\-\-symbols\fR]
            [\fB\-u\fR|\fB\-\-unwind\fR]
            [\fB\-D\fR|\fB\-\-use-dynamic\fR]
            [\fB\-C\fR|\fB\-\-demangle\fR]
            [\fB\-m\fR <number or name>|\fB\-\-disassemble=\fR<number or name>]
            [\fB\-M\fR <options>|\fB\-\-disassembler\-options=\fR<options>]
            [\fB\-\-verbose]
            \fIelffile\fR...

.SH DESCRIPTION
\fBcoredump-ng\fR displays information about ELF format core dumps. The options control what particular information to display.

\fIelffile...\fR are the ELF format core dump to be examined. 32-bit and 64-bit core dumps are supported.

.SH OPTIONS
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to coredump-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of coredump-ng and exit.
.IP "\fB\-I\fR" 4
.IX Item "-I"
.PD 0
.IP "\fB\-\-histogram\fR" 4
.IX Item "--histogram"
.PD
Display a histogram of bucket list lengths when displaying the contents of the symbol tables.
.IP "\fB\-S\fR" 4
.IX Item "-S"
.PD 0
.IP "\fB\-\-section\-headers\fR" 4
.IX Item "--section-headers"
.PD
Displays the information contained in the file's section headers.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\-info\fR" 4
.IX Item "--version-info"
.PD
Displays the contents of the version sections in the file.
.IP "\fB\-a\fR" 4
.IX Item "-a"
.PD 0
.IP "\fB\-\-all\fR" 4
.IX Item "--all"
.PD
Equivalent to specifying: \fB\-\-arch-specific\fR, \fB\-\-histogram\fR, \fB\-\-section-headers\fR, \fB\-\-version-info\fR, \fB\-\-dynamic\fR, \fB\-\-section-groups\fR, \fB\-\-file-header\fR, \fB\-\-program-headers\fR, \fB\-\-notes\fR, \fB\-\-relocs\fR, \fB\-\-symbols\fR and \fB\-\-unwind\fR.
.IP "\fB\-d\fR" 4
.IX Item "-d"
.PD 0
.IP "\fB\-\-dynamic\fR" 4
.IX Item "--dynamic"
Displays the contents of the file's dynamic section.
.IP "\fB\-e\fR" 4
.IX Item "-e"
.PD 0
.IP "\fB\-\-headers\fR" 4
.IX Item "--headers"
.PD
Display all the headers in the file.
Equivalent to specifying: \fB\-\-section-headers\fR, \fB\-\-file-header\fR and \fB\-\-program-headers\fR.
.IP "\fB\-g\fR" 4
.IX Item "-g"
.PD 0
.IP "\fB\-\-section\-groups\fR" 4
.IX Item "--section-groups"
.PD
Displays the information contained in the file's section groups.
.IP "\fB\-t\fR" 4
.IX Item "-t"
.PD 0
.IP "\fB\-\-section\-details\fR" 4
.IX Item "--section-details"
.PD
Displays the detailed section information. Implies \fB\-S\fR.
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-file\-header\fR" 4
.IX Item "--file-header"
.PD
Displays the information contained in the ELF header at the start of the file.
.IP "\fB\-l\fR" 4
.IX Item "-l"
.PD 0
.IP "\fB\-\-program\-headers\fR" 4
.IX Item "--program-headers"
.PD
Displays the information contained in the file's segment headers.
.IP "\fB\-n\fR" 4
.IX Item "-n"
.PD 0
.IP "\fB\-\-notes\fR" 4
.IX Item "--notes"
.PD
Displays the contents of the NOTE segments.
.IP "\fB\-r\fR" 4
.IX Item "-r"
.PD 0
.IP "\fB\-\-relocs\fR" 4
.IX Item "--relocs"
.PD
Displays the contents of the file's relocation section.
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD 0
.IP "\fB\-\-symbols\fR" 4
.IX Item "--symbols"
.PD
Displays the entries in symbol table section of the file.
.IP "\fB\-u\fR" 4
.IX Item "-u"
.PD 0
.IP "\fB\-\-unwind\fR" 4
.IX Item "--unwind"
.PD
Displays the contents of the file's unwind section.
.IP "\fB\-D\fR" 4
.IX Item "-D"
.PD 0
.IP "\fB\-\-use\-dynamic\fR" 4
.IX Item "--use-dynamic"
.PD
Use the dynamic section info when displaying symbols.
.IP "\fB\-C\fR" 4
.IX Item "-C"
.PD 0
.IP "\fB\-\-demangle\fR" 4
.IX Item "--demangle"
.PD
Decode (demangle) low-level symbol names into user-level names. This makes C++ function names readable. Different compilers have different mangling styles.
.IP "\fB\-m <number or name>\fR" 4
.IX Item "-m <number or name>"
.PD 0
.IP "\fB\-\-disassemble=<number or name>\fR" 4
.IX Item "--disassemble=<number or name>"
.PD
Display the assembler mnemonics for the machine instructions from the input file.
.IP "\fB\-M <options>\fR" 4
.IX Item "-M <options>"
.PD 0
.IP "\fB\-\-disassembler\-options=<options>\fR" 4
.IX Item "--disassembler-options=<options>"
.PD
Pass target specific information to the disassembler.
.RS 4
.ie n .IP """att""" 4
.IX Item "att"
.PD
Select the AT&T syntax mode.
.ie n .IP """intel""" 4
.IX Item "intel"
.PD
Select the Intel syntax mode.
.ie n .IP """arm32""" 4
.IX Item "arm32"
.PD 0
.ie n .IP """arm64""" 4
.IX Item "arm64"
.ie n .IP """i386""" 4
.IX Item "i386"
.ie n .IP """i8086""" 4
.IX Item "i8086"
.ie n .IP """x86-64""" 4
.IX Item "x86-64"
.ie n .IP """riscv32""" 4
.IX Item "riscv32"
.ie n .IP """riscv64""" 4
.IX Item "riscv64"
.ie n .IP """mips1""" 4
.IX Item "mips1"
.ie n .IP """mips2""" 4
.IX Item "mips2"
.ie n .IP """mips3""" 4
.IX Item "mips3"
.ie n .IP """mips4""" 4
.IX Item "mips4"
.ie n .IP """mips5""" 4
.IX Item "mips5"
.ie n .IP """mips32""" 4
.IX Item "mips32"
.ie n .IP """mips32r2""" 4
.IX Item "mips32r2"
.ie n .IP """mips32r6""" 4
.IX Item "mips32r6"
.ie n .IP """mips64""" 4
.IX Item "mips64"
.ie n .IP """mips64r2""" 4
.IX Item "mips64r2"
.ie n .IP """mips64r6""" 4
.IX Item "mips64r6"
.PD
Select disassembly for the given architecture.
.ie n .IP """amd64""" 4
.IX Item "amd64"
.PD 0
.ie n .IP """intel64""" 4
.IX Item "intel64"
.PD
Select between AMD64 ISA and Intel64 ISA.
.ie n .IP """addr16""" 4
.IX Item "addr16"
.PD 0
.ie n .IP """addr32""" 4
.IX Item "addr32"
.ie n .IP """addr64""" 4
.IX Item "addr64"
.ie n .IP """data16""" 4
.IX Item "data16"
.ie n .IP """data32""" 4
.IX Item "data32"
.PD
Specify the default address size and operand size.
.RE
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/coredump-ng.c
#include "buffer.h"
#include "printf.h"
#include "scripts.h"
#include "options.h"
#include "coredump.h"

#include "static/coredump-ng.ci"

#define THIS_NAME "coredump-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zCOREDESCRIPTION);
  n += usage_synopsis0(o, name, args);
  n += usage_synopsis2(o, name, zDISASSEMBLEARGS, zCOREELFARGS2, zCOREELFARGS3);
  n += usage_objname(o, name, "objfiles");
  n += printf_eol();
  n += usage_description(o, name, args, zELFDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options2(o, name, zDISASSEMBLEARGS, zCOREELFARGS2, zCOREELFARGS3);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_coredump(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zCOREELFARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  o->action = OPRPROGRAM_NOWARNINGS | OPRPROGRAM_COREDUMPDETAILED;

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 1024);
      MALLOCA(char, arg1, 1024);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, zCOREELFARGS3)) {
          imode_t ocdump = get_options2(o, zDISASSEMBLEARGS, arg1);
          if (0 == ocdump) {
            return odeath(o, THIS_NAME, arg1);
          }
          o->ocdump |= ocdump;
        } else if (0 == xstrcmp(arg0, "--disassemble")) {
          oinsertsname(o, ACT_DISASSEMBLE, arg1);
        } else {
          return odeath(o, THIS_NAME, arg0 + 2);
        }
      } else {
        imode_t action = get_options2(o, zCOREELFARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 2);
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0]) {
      if ( argv[i][0] == zCOREELFARGS2[0] && argv[i][1] == zCOREELFARGS2[1]  ) {
        imode_t ocdump = get_options2(o, zDISASSEMBLEARGS, argv[++i]);
        if (0 == ocdump) {
          return odeath(o, THIS_NAME, argv[i]);
        }
        o->ocdump |= ocdump;
      } else if (0 == xstrcmp(argv[i], "-m")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_DISASSEMBLE, argv[++i]);
      } else {
        imode_t action = get_options1(o, zCOREELFARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (0 == o->inpname0[0]) {
      xstrncpy(o->inpname0, argv[i], NELEMENTS(o->inpname0));
    } else if (0 == o->inpname1[0]) {
      xstrncpy(o->inpname1, argv[i], NELEMENTS(o->inpname1));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  o->ocdump |= get_ocdump(o, zCOREELFSWAP, o->action);

  if (MODE_ISANY(o->action, OPTPROGRAM_VERSION)) {
    return version0(o, THIS_NAME, zCOREELFARGS);
  }

  if (MODE_ISANY(o->action, OPTPROGRAM_HELP)) {
    return usage(o, THIS_NAME, zCOREELFARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_coredump(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname0[0]) {
      pbuffer_t p = bopen(o->inpname0);
      pbuffer_t q = bopen(o->inpname1);
      if (p) {
        if (isELFcore(p)) {
          x = coredump(p, q, o);
        } else {
          printf_e("'%s': invalid file format.", o->inpname0);
        }
      } else {
        printf_e("'%s': no such file.", o->inpname0);
      }

     bfree(p);
     bfree(q);
    }

    ofree(o);
  }

  return x;
}

